### ISSUE TITLE
Detect binding to variables for alternative patterns

### ISSUE THREAD

**Issue by tejasbubane** [rubocop/rubocop#13973]
## Is your feature request related to a problem? Please describe.

Binding to variables currently does not work for alternative patterns joined with `|`

```ruby
case {a: 1, b: 2}
in {a: } | Array
  "matched: #{a}"
else
  "not matched"
end
```

```
# SyntaxError (illegal variable in alternative pattern (a))
```

However, rubocop even with prism parser passes:

```
$ rubocop -V
1.73.2 (using Prism 1.2.0, rubocop-ast 1.38.1, analyzing as Ruby 3.4, running on ruby 3.4.2) [arm64-darwin24]
```

## Describe the solution you'd like

Rubocop should detect this error.

## Describe alternatives you've considered

It is a syntax error in CRuby so maybe Prism should detect it?

## Additional context

NA

**Comment by Earlopain** [comment]
Some "syntax errors" are only catched by ruby at runtime (or while compiling if you want to be more correct). You can check this with the `-c` flag:

```rb
$ ruby -c test.rb
Syntax OK
```

Only when you actually run it will it complain. I guess it would be too complicated for them to detect this purely in the parser. I encountered something else like that in the past, but I can't remember what it was now.

**Comment by tejasbubane** [comment]
Indeed the syntax check works fine. I also tested with `Prism.parse`.

> I guess it would be too complicated for them to detect this purely in the parser

So I guess it is difficult for rubocop as well?

**Comment by tejasbubane** [comment]
I see we have `InPatternNode` with `:match_alt` and `:match_var`. 

```
3.4.2 :003 > node.in_pattern_branches
 => 
[s(:in_pattern,
  s(:match_alt,
    s(:array_pattern,
      s(:match_var, :a)),
    s(:const, nil, :Array)), nil,
  s(:send, nil, :do_something))] 
```

So I guess it should be possible for rubocop to detect this. I will try to create a `Lint` cop.

**Comment by Earlopain** [comment]
Created prism issue https://github.com/ruby/prism/issues/3541
